### PR TITLE
[amazonechocontrol] last spoken text channel

### PR DESCRIPTION
**Pull request by Trinitus01**
changed: last voice command via customer history record
added: last spoken text channel

Signed-off-by: Tom Blum <trinitus01@googlemail.com>

### CODE REVIEW

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/AmazonEchoControlBindingConstants.java [hunk 1]
```diff
@@ -120,82 +121,44 @@
 
     // DeviceTypeIds to human readable description
     // originally found here: https://github.com/Apollon77/ioBroker.alexa2/blob/master/main.js
-    public static final Map<String, String> DEVICE_TYPES = Map.<String, String> ofEntries( //
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/AmazonEchoControlBindingConstants.java]: Please revert this change.

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/Connection.java [hunk 2]
```diff
@@ -1150,6 +1153,21 @@ public String getDeviceListJson() throws IOException, URISyntaxException, Interr
         return List.of();
     }
 
+    public List<CustomerHistoryRecord> getActivities(@Nullable Long startTime, @Nullable Long endTime) {
+        try {
+            String json = makeRequestAndReturnString(
+                    "https://" + amazonSite + "/alexa-privacy/apd/rvh/customer-history-records?startTime="
+                            + (startTime != null ? startTime : "") + "&endTime=" + (endTime != null ? endTime : "")
+                            // + "&recordType=VOICE_HISTORY"
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/Connection.java]: remove commented code

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java [hunk 3]
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
...
+        /*
+         * DEPRECATED
+         * connection.getActivities(10, pushActivity.timestamp).stream().filter(activity -> search.equals(activity.id))
+         * .findFirst()
+         * .ifPresent(currentActivity -> currentActivity.getSourceDeviceIds().stream()
+         * .map(sourceDeviceId -> findEchoHandlerBySerialNumber(sourceDeviceId.serialNumber))
+         * .filter(Objects::nonNull).forEach(echoHandler -> Objects.requireNonNull(echoHandler)
+         * .handlePushActivity(currentActivity)));
+         */
+
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java]: remove commented code. If it is needed later, it can be retrieved from the git history.

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java [hunk 4]
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
...
+        Long timestamp = pushActivity.timestamp;
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:788]:
```suggestion
        long timestamp = pushActivity.timestamp;
```

use primitives where possible

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java [hunk 5]
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
...
+        if (timestamp != null) {
+            Long startTimestamp = timestamp - 30000;
+            Long endTimestamp = timestamp + 30000;
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java]:
```suggestion
            long startTimestamp = timestamp - 30000;
            long endTimestamp = timestamp + 30000;
```

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java [hunk 6]
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
...
+            List<CustomerHistoryRecord> customerHistoryRecords = connection.getActivities(startTimestamp, endTimestamp);
+            for (CustomerHistoryRecord customerHistoryRecord : customerHistoryRecords) {
+                if (customerHistoryRecord != null) {
+                    String recordKey = customerHistoryRecord.recordKey;
+                    if (recordKey != null && search.equals(recordKey)) {
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:797]:
```suggestion
                    if (search.equals(recordKey)) {
```

No need to check for null: `search` is not null and .equals is always null-safe (for the second object)

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java [hunk 7]
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
...
+                        EchoHandler echoHandler = findEchoHandlerBySerialNumber(recordKey.split("#")[3]);
```
Comment (J-N-K) [inline bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java]: Is this safe? If recordKey.split returns less than four parts, this will result in an IndexOutOfBoundsException

**Review by J-N-K — CHANGES_REQUESTED** [review]
Thanks. Please add the new channel to README, too.

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java (hunk 4 above)
Comment (Trinitus01) [reply bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:788]: timestamp of pushActivity is a nullable Long so have to use Long and check for null.

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java (hunk 6 above)
Comment (Trinitus01) [reply bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:797]: The check for null is not because of the following equals-method, it is because of the use of recordKey inside the if statement.

File: bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java (hunk 7 above)
Comment (Trinitus01) [reply bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java]: It depends on the data of the response, so i will add a check.

**Review by J-N-K — APPROVED** [review]
